### DISCUSSION TITLE
Using IterableDataset with Torch DataLoader throws error.

### DISCUSSION THREAD

**Discussion by LeenaShekhar** [huggingface/datasets#2577] [Q&A]
Hi,

I have a general question regarding using Dataset in Streaming mode – Is IterableDataset not to be used with Pytorch DataLoader? I can use Dataset with the DataLoader without any issues (as is also mentioned in the examples), but I cannot do so with the former. I am quite new to the HF Dataset library so my apologies if this is already mentioned somewhere (I am still looking).

I get the following error, which makes sense because this is streaming mode, but I am unclear about how to design so that I can do batching then:
<code>
File “/data/leshekha/lib/HFDatasets/lib/python3.6/site-packages/torch/utils/data/sampler.py”, line 67, in iter
return iter(range(len(self.data_source)))
TypeError: object of type ‘IterableDataset’ has no len()
</code>

Any help is appreciated. Thank you. 

I have asked the same question here: https://discuss.huggingface.co/t/roadmap-timeline-for-dataset-streaming/6789/5

**Comment by changjonathanc** [comment]
PyTorch only recognize  `torch.utils.data.IterableDataset`.
You can use this:
```
torch_iter_dataset = hf_iter_dataset.with_format("torch")
```
Or you can customize with something like this:
```
import torch

class TorchIterableDataset(torch.utils.data.IterableDataset):
    def __init__(self, hf_iter_ds):
        super().__init__()
        self.ds = hf_iter_ds

    def __iter__(self):
        for ex in self.ds:
            yield ex
```

Reference:
https://github.com/huggingface/datasets/blob/c7228105da8fa981671eb82def118a59000d0c8c/src/datasets/iterable_dataset.py#L480-L494

**Reply by LeenaShekhar** [reply]
Thank you for the response. Yes, I was given a similar answer here: https://github.com/huggingface/datasets/issues/2583.
